File: language/runtime/android/kotlin/src/main/kotlin/dev/destack/runtime/android/module/background/RuntimeHostBackground.kt
package dev.destack.runtime.android.module.background

/**
 * The Android background scheduler availability state.
 */
public enum class RuntimeHostBackgroundStatus(
    /**
     * The raw bridge value for this enum case.
     */
    val rawValue: Int,
) {
    /**
     * The scheduler is unavailable on this host.
     */
    Unavailable(1),

    /**
     * The scheduler is available but restricted.
     */
    Restricted(2),

    /**
     * The scheduler is available.
     */
    Available(3),
}

/**
 * The Android background trigger class.
 */
public enum class RuntimeHostBackgroundTriggerKind(
    /**
     * The raw bridge value for this enum case.
     */
    val rawValue: Int,
) {
    /**
     * One opportunistic refresh task.
     */
    AppRefresh(1),

    /**
     * One longer processing task.
     */
    Processing(2),
}

/**
 * The Android background completion class.
 */
public enum class RuntimeHostBackgroundTaskResult(
    /**
     * The raw bridge value for this enum case.
     */
    val rawValue: Int,
) {
    /**
     * The task completed successfully.
     */
    Success(1),

    /**
     * The task should be retried.
     */
    Retry(2),

    /**
     * The task failed permanently.
     */
    Failure(3),
}

/**
 * The Android background network requirement.
 */
public enum class RuntimeHostBackgroundNetworkRequirement(
    /**
     * The raw bridge value for this enum case.
     */
    val rawValue: Int,
) {
    /**
     * No network route is required.
     */
    None(1),

    /**
     * One network route is required.
     */
    Connected(2),

    /**
     * One unmetered network route is required.
     */
    Unmetered(3),
}

/**
 * The Android background conflict policy.
 */
public enum class RuntimeHostBackgroundConflictPolicy(
    /**
     * The raw bridge value for this enum case.
     */
    val rawValue: Int,
) {
    /**
     * Replace one existing registration.
     */
    Replace(1),

    /**
     * Keep one existing registration.
     */
    Keep(2),
}

/**
 * The Android background schedule class.
 */
public enum class RuntimeHostBackgroundTaskScheduleKind(
    /**
     * The raw bridge value for this enum case.
     */
    val rawValue: Int,
) {
    /**
     * One single future execution.
     */
    Once(1),

    /**
     * One recurring execution stream.
     */
    Recurring(2),
}

/**
 * One Android background task schedule payload.
 */
public data class RuntimeHostBackgroundTaskSchedule(
    /**
     * The schedule class.
     */
    val kind: RuntimeHostBackgroundTaskScheduleKind,

    /**
     * The earliest execution target in UTC nanoseconds when provided.
     */
    val earliestBeginUnixNs: Long?,

    /**
     * The repeat interval in nanoseconds for recurring schedules.
     */
    val repeatIntervalNs: Long? = null,
)

/**
 * One Android background task registration payload.
 */
public data class RuntimeHostBackgroundTaskOptions(
    /**
     * The stable task identifier.
     */
    val identifier: String,

    /**
     * The trigger class.
     */
    val trigger: RuntimeHostBackgroundTriggerKind,

    /**
     * The requested schedule payload.
     */
    val schedule: RuntimeHostBackgroundTaskSchedule,

    /**
     * The requested network requirement.
     */
    val network: RuntimeHostBackgroundNetworkRequirement,

    /**
     * Whether charging power is required.
     */
    val requiresCharging: Boolean,

    /**
     * Whether device idle mode is required.
     */
    val requiresIdle: Boolean,

    /**
     * The conflict policy for one existing registration.
     */
    val conflictPolicy: RuntimeHostBackgroundConflictPolicy,
)

/**
 * One Android background status response.
 */
public data class RuntimeHostBackgroundStatusResponse(
    /**
     * The host status code.
     */
    val status: Int,

    /**
     * The scheduler status when available.
     */
    val schedulerStatus: RuntimeHostBackgroundStatus? = null,
)

/**
 * One Android background task-list response.
 */
public data class RuntimeHostBackgroundListResponse(
    /**
     * The host status code.
     */
    val status: Int,

    /**
     * The returned task descriptors.
     */
    val descriptors: List<RuntimeHostBackgroundTaskDescriptor> = emptyList(),
)

/**
 * One Android background unregister request.
 */
public data class RuntimeHostBackgroundUnregisterRequest(
    /**
     * The stable task identifier.
     */
    val identifier: String,
)

/**
 * One Android background trigger-test request.
 */
public data class RuntimeHostBackgroundTriggerTestRequest(
    /**
     * The stable task identifier.
     */
    val identifier: String,
)

/**
 * One Android background trigger-test response.
 */
public data class RuntimeHostBackgroundTriggerTestResponse(
    /**
     * The host status code.
     */
    val status: Int,

    /**
     * Whether the host triggered one task execution.
     */
    val isTriggered: Boolean = false,
)

/**
 * One Android background complete request.
 */
public data class RuntimeHostBackgroundCompleteRequest(
    /**
     * The stable execution identifier.
     */
    val executionId: String,

    /**
     * The completion result.
     */
    val result: RuntimeHostBackgroundTaskResult,
)

/**
 * One Android background task descriptor payload.
 */
public data class RuntimeHostBackgroundTaskDescriptor(
    /**
     * The stable task identifier.
     */
    val identifier: String,

    /**
     * The trigger class.
     */
    val trigger: RuntimeHostBackgroundTriggerKind,

    /**
     * The effective schedule payload.
     */
    val schedule: RuntimeHostBackgroundTaskSchedule,

    /**
     * The effective network requirement.
     */
    val network: RuntimeHostBackgroundNetworkRequirement,

    /**
     * Whether charging power is required.
     */
    val requiresCharging: Boolean,

    /**
     * Whether device idle mode is required.
     */
    val requiresIdle: Boolean,

    /**
     * The conflict policy for one existing registration.
     */
    val conflictPolicy: RuntimeHostBackgroundConflictPolicy,
) {
    /**
     * Return the flattened schedule kind for the Android bridge.
     */
    val scheduleKind: RuntimeHostBackgroundTaskScheduleKind
        get() = schedule.kind

    /**
     * Return whether this descriptor uses one recurring schedule.
     */
    val isRecurring: Boolean
        get() = schedule.kind == RuntimeHostBackgroundTaskScheduleKind.Recurring

    /**
     * Return whether one earliest execution target is present for the Android bridge.
     */
    val hasEarliestBeginUnixNs: Boolean
        get() = schedule.earliestBeginUnixNs != null

    /**
     * Return the flattened earliest execution target for the Android bridge.
     */
    val earliestBeginUnixNs: Long
        get() = schedule.earliestBeginUnixNs ?: 0L

    /**
     * Return whether one repeat interval is present for the Android bridge.
     */
    val hasRepeatIntervalNs: Boolean
        get() = schedule.repeatIntervalNs != null

    /**
     * Return the flattened repeat interval for the Android bridge.
     */
    val repeatIntervalNs: Long
        get() = schedule.repeatIntervalNs ?: 0L
}

/**
 * One Android background event metadata payload.
 */
public data class RuntimeHostBackgroundEventMetadata(
    /**
     * The monotonic event timestamp in nanoseconds.
     */
    val timestampNs: Long,

    /**
     * The monotonic event sequence number.
     */
    val sequence: Long,

    /**
     * The stable task identifier.
     */
    val identifier: String,

    /**
     * The stable execution identifier.
     */
    val executionId: String,

    /**
     * The execution deadline in UTC nanoseconds.
     */
    val deadlineUnixNs: Long,
)

/**
 * The Android background event kind.
 */
public enum class RuntimeHostBackgroundEventKind(
    /**
     * The raw bridge value for this enum case.
     */
    val rawValue: Int,
) {
    /**
     * One task became ready to execute.
     */
    TaskReady(1),

    /**
     * One task execution expired.
     */
    TaskExpired(2),
}

/**
 * One Android background ingress event delivered into one runtime session.
 */
public data class RuntimeHostBackgroundEvent(
    /**
     * The event kind.
     */
    val kind: RuntimeHostBackgroundEventKind,

    /**
     * The shared event metadata.
     */
    val metadata: RuntimeHostBackgroundEventMetadata,
)

File: language/runtime/apple/core/Sources/RuntimeHostAppleCore/Module/Background/RuntimeHostBackground.swift
import Foundation

/// The Apple background scheduler availability state.
public enum RuntimeHostBackgroundStatus: UInt32, Sendable, Hashable, Codable {
  /// The scheduler is unavailable on this host.
  case unavailable = 1
  /// The scheduler is available but restricted.
  case restricted = 2
  /// The scheduler is available.
  case available = 3
}

/// The Apple background trigger class.
public enum RuntimeHostBackgroundTriggerKind: UInt32, Sendable, Hashable, Codable {
  /// One opportunistic refresh task.
  case appRefresh = 1
  /// One longer processing task.
  case processing = 2
}

/// The Apple background network requirement.
public enum RuntimeHostBackgroundNetworkRequirement: UInt32, Sendable, Hashable, Codable {
  /// No network route is required.
  case none = 1
  /// One network route is required.
  case connected = 2
  /// One unmetered network route is required.
  case unmetered = 3
}

/// The Apple background conflict policy.
public enum RuntimeHostBackgroundConflictPolicy: UInt32, Sendable, Hashable, Codable {
  /// Replace one existing registration.
  case replace = 1
  /// Keep one existing registration.
  case keep = 2
}

/// The Apple background schedule class.
public enum RuntimeHostBackgroundTaskScheduleKind: UInt32, Sendable, Hashable, Codable {
  /// One single future execution.
  case once = 1
  /// One recurring execution stream.
  case recurring = 2
}

/// The Apple background completion class.
public enum RuntimeHostBackgroundTaskResult: UInt32, Sendable, Hashable, Codable {
  /// The task completed successfully.
  case success = 1
  /// The task should be retried.
  case retry = 2
  /// The task failed permanently.
  case failure = 3
}

/// One Apple background task schedule payload.
public struct RuntimeHostBackgroundTaskSchedule: Sendable, Hashable, Codable {
  /// The schedule class.
  public let kind: RuntimeHostBackgroundTaskScheduleKind
  /// The earliest execution target in UTC nanoseconds when provided.
  public let earliestBeginUnixNs: UInt64?
  /// The repeat interval in nanoseconds for recurring schedules.
  public let repeatIntervalNs: UInt64?

  /// Create one background task schedule payload.
  public init(
    kind: RuntimeHostBackgroundTaskScheduleKind,
    earliestBeginUnixNs: UInt64?,
    repeatIntervalNs: UInt64? = nil
  ) {
    self.kind = kind
    self.earliestBeginUnixNs = earliestBeginUnixNs
    self.repeatIntervalNs = repeatIntervalNs
  }

  /// Return whether this schedule uses recurring execution.
  public var isRecurring: Bool {
    kind == .recurring
  }
}

/// One Apple background task registration payload.
public struct RuntimeHostBackgroundTaskOptions: Sendable, Hashable, Codable {
  /// The stable task identifier.
  public let identifier: String
  /// The trigger class.
  public let trigger: RuntimeHostBackgroundTriggerKind
  /// The requested schedule payload.
  public let schedule: RuntimeHostBackgroundTaskSchedule
  /// The requested network requirement.
  public let network: RuntimeHostBackgroundNetworkRequirement
  /// Whether charging power is required.
  public let requiresCharging: Bool
  /// Whether idle mode is required.
  public let requiresIdle: Bool
  /// The registration conflict policy.
  public let conflictPolicy: RuntimeHostBackgroundConflictPolicy

  /// Create one background task options payload.
  public init(
    identifier: String,
    trigger: RuntimeHostBackgroundTriggerKind,
    schedule: RuntimeHostBackgroundTaskSchedule,
    network: RuntimeHostBackgroundNetworkRequirement,
    requiresCharging: Bool,
    requiresIdle: Bool,
    conflictPolicy: RuntimeHostBackgroundConflictPolicy
  ) {
    self.identifier = identifier
    self.trigger = trigger
    self.schedule = schedule
    self.network = network
    self.requiresCharging = requiresCharging
    self.requiresIdle = requiresIdle
    self.conflictPolicy = conflictPolicy
  }
}

/// One Apple background status response.
public struct RuntimeHostBackgroundStatusResponse: Sendable, Hashable, Codable {
  /// The host status code.
  public let status: UInt32
  /// The scheduler status when available.
  public let schedulerStatus: RuntimeHostBackgroundStatus?

  /// Create one background status response.
  public init(
    status: UInt32,
    schedulerStatus: RuntimeHostBackgroundStatus? = nil
  ) {
    self.status = status
    self.schedulerStatus = schedulerStatus
  }
}

/// One Apple background task descriptor payload.
public struct RuntimeHostBackgroundTaskDescriptor: Sendable, Hashable, Codable {
  /// The stable task identifier.
  public let identifier: String
  /// The trigger class.
  public let trigger: RuntimeHostBackgroundTriggerKind
  /// The effective schedule payload.
  public let schedule: RuntimeHostBackgroundTaskSchedule
  /// The effective network requirement.
  public let network: RuntimeHostBackgroundNetworkRequirement
  /// Whether charging power is required.
  public let requiresCharging: Bool
  /// Whether idle mode is required.
  public let requiresIdle: Bool
  /// The registration conflict policy.
  public let conflictPolicy: RuntimeHostBackgroundConflictPolicy

  /// Create one background task descriptor.
  public init(
    identifier: String,
    trigger: RuntimeHostBackgroundTriggerKind,
    schedule: RuntimeHostBackgroundTaskSchedule,
    network: RuntimeHostBackgroundNetworkRequirement,
    requiresCharging: Bool,
    requiresIdle: Bool,
    conflictPolicy: RuntimeHostBackgroundConflictPolicy
  ) {
    self.identifier = identifier
    self.trigger = trigger
    self.schedule = schedule
    self.network = network
    self.requiresCharging = requiresCharging
    self.requiresIdle = requiresIdle
    self.conflictPolicy = conflictPolicy
  }
}

/// One Apple background task-list response.
public struct RuntimeHostBackgroundListResponse: Sendable, Hashable, Codable {
  /// The host status code.
  public let status: UInt32
  /// The returned task descriptors.
  public let descriptors: [RuntimeHostBackgroundTaskDescriptor]

  /// Create one background task-list response.
  public init(
    status: UInt32,
    descriptors: [RuntimeHostBackgroundTaskDescriptor] = []
  ) {
    self.status = status
    self.descriptors = descriptors
  }
}

/// One Apple background event metadata payload.
public struct RuntimeHostBackgroundEventMetadata: Sendable, Hashable, Codable {
  /// The monotonic event timestamp in nanoseconds.
  public let timestampNs: UInt64
  /// The monotonic event sequence number.
  public let sequence: UInt64
  /// The stable task identifier.
  public let identifier: String
  /// The stable execution identifier.
  public let executionID: String
  /// The execution deadline in UTC nanoseconds.
  public let deadlineUnixNs: UInt64

  /// Create one background event metadata payload.
  public init(
    timestampNs: UInt64,
    sequence: UInt64,
    identifier: String,
    executionID: String,
    deadlineUnixNs: UInt64
  ) {
    self.timestampNs = timestampNs
    self.sequence = sequence
    self.identifier = identifier
    self.executionID = executionID
    self.deadlineUnixNs = deadlineUnixNs
  }
}

/// The Apple background event kind.
public enum RuntimeHostBackgroundEventKind: UInt32, Sendable, Codable {
  /// One task became ready to execute.
  case taskReady = 1
  /// One task execution expired.
  case taskExpired = 2
}

/// One Apple background ingress event delivered into one runtime session.
public struct RuntimeHostBackgroundEvent: Sendable, Hashable, Codable {
  /// The event kind.
  public let kind: RuntimeHostBackgroundEventKind
  /// The shared event metadata.
  public let metadata: RuntimeHostBackgroundEventMetadata

  /// Create one background event.
  public init(
    kind: RuntimeHostBackgroundEventKind,
    metadata: RuntimeHostBackgroundEventMetadata
  ) {
    self.kind = kind
    self.metadata = metadata
  }
}

/// One background unregister request.
public struct RuntimeHostBackgroundUnregisterRequest: Sendable, Hashable, Codable {
  /// The stable task identifier.
  public let identifier: String

  /// Create one background unregister request.
  public init(identifier: String) {
    self.identifier = identifier
  }
}

/// One background trigger-test request.
public struct RuntimeHostBackgroundTriggerTestRequest: Sendable, Hashable, Codable {
  /// The stable task identifier.
  public let identifier: String

  /// Create one background trigger-test request.
  public init(identifier: String) {
    self.identifier = identifier
  }
}

/// One Apple background trigger-test response.
public struct RuntimeHostBackgroundTriggerTestResponse: Sendable, Hashable, Codable {
  /// The host status code.
  public let status: UInt32
  /// Whether the host triggered one task execution.
  public let isTriggered: Bool

  /// Create one background trigger-test response.
  public init(
    status: UInt32,
    isTriggered: Bool = false
  ) {
    self.status = status
    self.isTriggered = isTriggered
  }
}

/// One background complete request.
public struct RuntimeHostBackgroundCompleteRequest: Sendable, Hashable, Codable {
  /// The stable execution identifier.
  public let executionID: String
  /// The completion result.
  public let result: RuntimeHostBackgroundTaskResult

  /// Create one background complete request.
  public init(
    executionID: String,
    result: RuntimeHostBackgroundTaskResult
  ) {
    self.executionID = executionID
    self.result = result
  }
}
